### ISSUE TITLE
Rule proposal: prefer-structured-clone

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8940]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Report on old ways of cloning objects when the newer `structuredClone` would be preferable.

### Fail Cases

```TypeScript
var clonedObject = JSON.parse(JSON.stringify(thingToBeCloned));
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// possibly could be controlled by an option { "reportShallowClone": "always" | "never" | "shallowObjects" } 
// the "shallowObjects" option would require type information to ensure that all the properties
// of `thingToBeCloned` are primitives, rather than objects
var clonedObject = { ...thingToBeCloned };
                   ^^^^^^^^^^^^^^^^^^^^^^
// same as previous
var clonedObject = Object.assign({}, thingToBeCloned);
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

// provide suggestion for all above examples to 
var clonedObject = structuredClone(thingToBeCloned);
```

### Pass Cases

```TypeScript
var clonedObject = structuredClone(thingToBeCloned);

// ignore anything with replacer/reviver arguments
var clonedObject = JSON.parse(JSON.stringify(thingToBeCloned, f), g);

var justAnObjectSpread = { ...object1, ...object2 };
var sameAsAbove = Object.assign({}, object1, object2);

var objectAssignOnNonObjectLiteral = Object.assign(object1, object2);
```

### Additional Info

This rule was (understandably) rejected from eslint core (see more at https://github.com/eslint/eslint/issues/18354). Had it been accepted in eslint core, I would have proposed an extension rule here for the type-information-requiring "shallowObjects" option.

I realize it may be rejected here for similar reasons, but figure it's worth a shot. I think the rule has plenty of value as a TS-specific rule anyway, since a TS version of it can whether an object is shallow, as previously mentioned, and we can also put more validation on the JSON cloning as well to determine whether an object's declared shape is JSON-cloneable. 

Additionally, whereas `prefer-structured-clone` may not be appropriate for the ESLint core, due to `structuredClone` not being technically part of the ECMA standard, it _is_ available in TS's type definitions, due to it being widely available in the real world (including outside browsers, i.e. Node >= 17 and Deno support it).

**Comment by kirkwaiblinger** [comment]
For fun, an example of code in the typescript-eslint repo that would flag:
https://github.com/typescript-eslint/typescript-eslint/blob/f81314731cccb779423e2580a805eff3efff8564/packages/eslint-plugin/src/util/getFunctionHeadLoc.ts#L199-L202

**Comment by Josh-Cena** [comment]
I'm not big about this. As you realized it's not ES standard and not every runtime has it. We have never prioritized a web feature before.

**Comment by Josh-Cena** [comment]
If you want to detect multi-level objects and report them being shallowly cloned, I'm totally fine, but the rule should just be called `no-partial-clone` or something. On the other hand I don't think JSON round-tripping is bad for simple cases.
